### ISSUE TITLE
The alternatives module is erroneously reporting "changed" status.

### ISSUE THREAD

**Issue by yeroc** [ansible/ansible-modules-extras#12]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.8 (devel ba4693051b) last updated 2014/08/06 15:37:06 (GMT -600)
##### Environment:

CentOS 6
##### Summary:

The alternatives module is reporting "changed" status when re-running a task that has already been successfully applied.
##### Steps To Reproduce:

Given a play:

```

---
- hosts: all
  sudo: yes
  vars:
    jdk_dir: /usr/java/jdk1.7.0_55

  tasks:
  - name: Configure Java alternatives
    alternatives: name={{ item.name }} link={{ item.link }} path={{ item.path }}
    with_items:
    - { name: jar, link: /usr/bin/jar, path: "{{ jdk_dir }}/bin/jar" }
    - { name: jps, link: /usr/bin/jps, path: "{{ jdk_dir }}/bin/jps" }
    - { name: java, link: /usr/bin/java, path: "{{ jdk_dir }}/bin/java" }
    - { name: jmap, link: /usr/bin/jmap, path: "{{ jdk_dir }}/bin/jmap" }
    - { name: javac, link: /usr/bin/javac, path: "{{ jdk_dir }}/bin/javac" }
    - { name: javaws, link: /usr/bin/javaws, path: "{{ jdk_dir }}/bin/javaws" }
    - { name: jstack, link: /usr/bin/jstack, path: "{{ jdk_dir }}/bin/jstack" }
    - { name: jvisualvm, link: /usr/bin/jvisualvm, path: "{{ jdk_dir }}/bin/jvisualvm" }
```
##### Expected Results:

The first time the task is run the status is "changed" as expected.  We expect subsequent runs with the identical path specified to indicate an "ok" status.
##### Actual Results:

When running the task subsequently the status is actually reporting "changed".

Note: This ticket was previously reported as [ansible #8613](https://github.com/ansible/ansible/issues/8613).

**Comment by MichaelPereira** [comment]
I am getting a different error when using this same module on Centos 6 (see #46), what python version are you using?

**Comment by yeroc** [comment]
I was using a newer version of ansible than yours so maybe that explains the difference?  It's possible the issue you've reported has been fixed in the v1.8 series?  We're just using the standard version of python that ships with CentOS 6.5.

**Comment by bcoca** [comment]
so i cannot reproduce this in current devel, can you give it a try?

**Comment by Mark2009** [comment]
I ran into this on 1.8.4, Ubuntu 14.04.2 LTS as the ansible master, invoking alternatives against RHEL 6.6.

**Comment by bcoca** [comment]
@Mark2009 can you test against current development? there is a patch but i cannot reproduce this issue so i cannot verify it as fixed.

**Comment by yeroc** [comment]
@bcoca I just tried to verify this but things are now blowing up with other errors.  Not sure if the alternatives module is currently broken in dev?  

I tested using:

```
ansible-playbook 1.9 (devel c81d981164) last updated 2015/02/25 10:50:42 (GMT -600)
  lib/ansible/modules/core: (detached HEAD 706be79f8f) last updated 2015/02/25 10:50:55 (GMT -600)
  lib/ansible/modules/extras: (detached HEAD 2ffd7235e5) last updated 2015/02/25 10:50:58 (GMT -600)
  v2/ansible/modules/core: (detached HEAD 34784b7a61) last updated 2015/02/25 10:51:02 (GMT -600)
  v2/ansible/modules/extras: (detached HEAD 312d59061c) last updated 2015/02/25 10:51:05 (GMT -600)
```

Partial output:

```
<192.168.33.10> path=/usr/java/jdk1.7.0_55/bin/jvisualvm link=/usr/bin/jvisualvm name=jvisualvm
<192.168.33.10> 
<192.168.33.10> 
<192.168.33.10> 'sudo -k && sudo -H -S -p "[sudo via ansible, key=dfaamhnreixhkszwbotflkpeiuzzqwbv] password: " -u root /bin/sh -c '"'"'echo SUDO-SUCCESS-dfaamhnreixhkszwbotflkpeiuzzqwbv; LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /tmp/.ansible/ansible-tmp-1424888740.45-89617031853269/alternatives; rm -rf /tmp/.ansible/ansible-tmp-1424888740.45-89617031853269/ >/dev/null 2>&1'"'"''
failed: [dbload-centos] => (item={'path': u'/usr/java/jdk1.7.0_55/bin/jvisualvm', 'link': '/usr/bin/jvisualvm', 'name': 'jvisualvm'}) => {"failed": true, "item": {"link": "/usr/bin/jvisualvm", "name": "jvisualvm", "path": "/usr/java/jdk1.7.0_55/bin/jvisualvm"}, "parsed": false}
SUDO-SUCCESS-dfaamhnreixhkszwbotflkpeiuzzqwbv
Traceback (most recent call last):
  File "/tmp/.ansible/ansible-tmp-1424888740.45-89617031853269/alternatives", line 1756, in <module>
    main()
  File "/tmp/.ansible/ansible-tmp-1424888740.45-89617031853269/alternatives", line 126, in main
    line_name, line_mode, line_path = line.strip().split("\t")
ValueError: need more than 1 value to unpack
```

Target host is running CentOS 6.4.

**Comment by bcoca** [comment]
@yeroc I have not been able to reproduce that, can you provide a minimal playbook?

**Comment by yeroc** [comment]
@bcoca [Here's a gist](https://gist.github.com/yeroc/2db8dca5eb31dad7a046) containing Vagrantfile, ansible-inventory & self-contained ansible-play-alternatives-test.yml (minimal playbook).  You should be able to drop those three files into a directory and run `vagrant up` to reproduce.

**Comment by moorereason** [comment]
I'm seeing the original issue (always reports changed) with Ansible 1.8.2 on RH6 using the following task:

``` yaml
- name: install | update java alternatives
  alternatives:
    name: java
    path: "/opt/jdk{{ oraclejdk_version.major }}/jre/bin/java"
    link: /usr/bin/java
```

Here's the problem:

The module [correctly detects RedHat](https://github.com/ansible/ansible-modules-extras/blob/c7e3aee0f018c019860cf6f17ffa9be9b12b7155/system/alternatives.py#L118) since the initial command (with `--query`) returns 2.  The problem is that RH6 does not have a `--list` option, so current_path is never set.

```
sh# update-alternatives --list
alternatives version 1.3.49.3 - Copyright (C) 2001 Red Hat, Inc.
This may be freely redistributed under the terms of the GNU Public License.

usage: alternatives --install <link> <name> <path> <priority>
                    [--initscript <service>]
                    [--slave <link> <name> <path>]*
       alternatives --remove <name> <path>
       alternatives --auto <name>
       alternatives --config <name>
       alternatives --display <name>
       alternatives --set <name> <path>

common options: --verbose --test --help --usage --version
                --altdir <directory> --admindir <directory>
```

For RH6, the module should be using `--display`.  Example:

```
sh# update-alternatives --display mail
mail - status is manual.
 link currently points to /bin/mailx
/bin/mailx - priority 90
 slave mail-Mail: /bin/mailx
 slave mail-Mail-man: /usr/share/man/man1/mailx.1.gz
 slave mail-mail-man: /usr/share/man/man1/mailx.1.gz
Current `best' version is /bin/mailx.
```

The current value is the second line: `link currently points to <path>`.  RH7 has a `--display` option exactly like RH6, so you could just use `--display` if you wanted to avoid having different logic for RH6/7.

**Comment by DavidWittman** [comment]
@moorereason I took your suggestions into account when pushing fixes to the alternatives module in #313. If you don't mind trying them out I'd greatly appreciate your feedback.
